test/dynamo/test_install_params_as_graph_attr.py

```diff
@@ -320,6 +333,152 @@ def test_linear_explicit(
+
+        def test_fn(x: torch.Tensor) -> torch.Tensor:
+            return net(x)
+
+        self.check_export_matches_expectation(test_fn, 1, (tensor,))
+
+    def test_transformer(self) -> None:
+        transformer = torch.nn.Transformer(d_model=64).eval()
+        src = torch.rand(10, 32, 64)
+        tgt = torch.rand(20, 32, 64)
+
+        self.check_export_matches_expectation(transformer, 2, (src, tgt))
+
+        def test_fn(src: torch.Tensor, tgt: torch.Tensor) -> torch.Tensor:
+            return transformer(src, tgt)
+
+        self.check_export_matches_expectation(test_fn, 2, (src, tgt))
+
+    def test_optimizing_params_in_input(self) -> None:
+        param = torch.nn.Parameter(torch.randn(1, 5))
+        net = SimpleLinearModule()
+
+        def test_fn(x: torch.Tensor) -> torch.Tensor:
+            return net(x)
+
+        self.check_export_matches_expectation(net, 1, (param,))
+        self.check_export_matches_expectation(test_fn, 1, (param,))
+
+        x = torch.randn(1, 5)
+
+        def test_fn2(x: torch.Tensor, param: torch.nn.Parameter) -> torch.Tensor:
+            return net(x) + param
+
+        # net gets installed, param does not here
+        self.check_export_matches_expectation(test_fn2, 2, (x, param))
+
+        def test_fn3(
+            x: torch.Tensor, list_params: list[torch.nn.Parameter]
+        ) -> torch.Tensor:
+            return net(x) + sum(list_params)
+
+        # list_params should not be installed or inlined here
+        self.check_export_matches_expectation(test_fn3, 2, (x, [param, param]))
+
+    def test_optimizing_buffer_in_input(self) -> None:
+        buf = torch.nn.Buffer(data=torch.ones((1, 5)))
+        net = SimpleLinearModule()
+
+        def test_fn(x: torch.Tensor) -> torch.Tensor:
+            return net(x)
+
+        self.check_export_matches_expectation(net, 1, (buf,))
+        self.check_export_matches_expectation(test_fn, 1, (buf,))
+
+        x = torch.randn(1, 5)
+
+        def test_fn2(x: torch.Tensor, buf: torch.nn.Buffer) -> torch.Tensor:
+            return net(x) + buf
+
+        # net gets installed, buf does not here
+        self.check_export_matches_expectation(test_fn2, 2, (x, buf))
+
+    def test_optimizing_buffer_and_param_in_input(self) -> None:
+        param = torch.nn.Parameter(torch.randn(5, 1))
+        buf = torch.nn.Buffer(data=torch.ones((1, 1)))
+        x = torch.randn(1, 5)
+
+        def test_linear_explicit(
+            x: torch.Tensor, a: torch.Tensor, b: torch.Tensor
+        ) -> torch.Tensor:
+            return a * x + b
+
+        # Now, param and buf are input so should not be inlined
+        self.check_export_matches_expectation(test_linear_explicit, 3, (x, param, buf))
+
+    # NOTE: Previously, this was not supported, because modifying state
+    # but with these changes this should be enabled
+    @unittest.expectedFailure  # TODO: Investigate how to fix
+    @torch._dynamo.config.patch(inline_inbuilt_nn_modules=True)
```

Comment:
Why doesn't it work?

Reply:
If I remove the assert inside the `symbolic_convert`, this will pass; that code path will still trigger though because here, `self` is realized as a `NNModuleVariable` and so modifying `self.a` is still trying to call `SET_ATTR` on an `NNModuleVariable` with a non `ConstantVariable`; but after removing that assert this will pass

See [output_graph.py::L858](https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/torch/_dynamo/output_graph.py#L858) for where this variable is instantiated,[ symbolic_convert.py::L2318](https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/torch/_dynamo/symbolic_convert.py#L2318) for the assertion error

Reply:
Update: I had the boolean condition in mutation_guard.py as `not is_export or not config.install_params_as_graph_attr`; upon inspection this is wrong, as the `is_dynamic_nn_module` will then return False when we want to install params; as a result, we wouldn't instantiate `self` as the right type 

Fixing this to `not is_export or config.install_params_as_graph_attr` resolves this issue as we correctly instantiate a UnspecializedNNModule and mark for mutation, but this test will still fail, as `a` is not a parameter or buffer; 

Therefore, we also need to modify the condition with which we install tensors in `builder.py` (which is now done) to install any tensor not explicitly passed in the input